src/server/frontend_wayland/wlr_screencopy_v1.cpp

```diff
@@ -154,29 +298,145 @@ void mf::WlrScreencopyManagerV1::capture_output_region(
     (void)overlay_cursor;
     auto const extents = OutputGlobal::from_or_throw(output).current_config().extents();
     auto const intersection = geom::Rectangle{{x, y}, {width, height}}.intersection_with(extents);
-    new WlrScreencopyFrameV1{frame, ctx, intersection};
+    new WlrScreencopyFrameV1{frame, this, ctx, {intersection, output}};
+}
+
+mf::WlrScreencopyManagerV1::CaptureArea::CaptureArea(FrameParams const& params)
+    : params{params},
+      damage_amount{DamageAmount::none}
+{
+}
+
+mf::WlrScreencopyManagerV1::CaptureArea::~CaptureArea()
+{
+
+}
+
+void mf::WlrScreencopyManagerV1::CaptureArea::apply_damage(std::optional<geom::Rectangle> const& damage)
+{
+    if (damage && damage_amount != DamageAmount::full)
+    {
+        auto const intersection = intersection_of(damage.value(), params.area);
+        if (intersection.size == geom::Size{})
+        {
+            return;
+        }
+
+        if (damage_amount == DamageAmount::partial)
+        {
+            damage_rect = geom::Rectangles{damage_rect, intersection}.bounding_rectangle();
+        }
+        else // damage_amount == DamageAmount::none
+        {
+            damage_amount = DamageAmount::partial;
+            damage_rect = intersection;
+        }
+    }
+    else
+    {
+        // If damage amount is already full, or if damage is nullopt (and thus full damage should be applied)
+        damage_amount = DamageAmount::full;
+    }
+
+    if (pending_frame)
```

Comment:
This check is repeated in `capture_frame()` (which suggests it is NOT a precondition) and at every call site (which suggests it IS a precondition).

Maybe it should be called `capture_any_pending_frame()` (and omit the checks at call sites)